### PR TITLE
Added Boyer Moore in java and AVL Trees in Cpp

### PR DESCRIPTION
**Pull request by ajaynair710**
Fixes #

#### Checklist:
- [ ] 4 space indentation.
- [ ] Coding conventions are followed.
- [ ] Input is taken dynamically.
- [ ] Sample Input / Output is added at the end of file.
- [ ] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- 
- 

#### Languages Used:
- 
- 

#### Files Added:
- 
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by ajaynair710** [comment]
Added #1829 
Added #1830

File: AVL_Tree/AVL_Tree.cpp [hunk 1]
```diff
@@ -0,0 +1,553 @@
+#include<iostream>
+
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove this line

File: AVL_Tree/AVL_Tree.cpp [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+#include<cstdio>
+
+#include<sstream>
+
+#include<algorithm>
+
+#define pow2(n) (1 << (n))
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the blank lines

File: AVL_Tree/AVL_Tree.cpp [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+
+using namespace std;
+
+ 
+
+/*
+
+ * Node Declaration
+
+ */
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the blank lines

File: AVL_Tree/AVL_Tree.cpp [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+struct avl_node
+
+{
+
+    int data;
+
+    struct avl_node *left;
+
+    struct avl_node *right;
+
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the blank lines here too. it's better if you provide some comments

File: AVL_Tree/AVL_Tree.cpp [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+}*root;
+
+ 
+
+/*
+
+ * Class Declaration
+
+ */
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove blank line

File: AVL_Tree/AVL_Tree.cpp [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+
+class avlTree
+
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: here too

File: AVL_Tree/AVL_Tree.cpp [hunk 7]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
         root->left = insert(root->left, value);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 8]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
 
         root = balance (root);
-
     }
-
     else if (value >= root->data)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 9]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
     {
-
         root->right = insert(root->right, value);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 10]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
- */
-
 void avlTree::display(avl_node *ptr, int level)
-
 {
-
     int i;
-
     if (ptr!=NULL)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep space before and after !=

File: AVL_Tree/AVL_Tree.cpp [hunk 11]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
     {
-
         display(ptr->right, level + 1);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 12]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         printf("\n");
-
         if (ptr == root)
-
         cout<<"Root -> ";
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after <<

File: AVL_Tree/AVL_Tree.cpp [hunk 13]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         for (i = 0; i < level && ptr != root; i++)
-
             cout<<"        ";
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after <<

File: AVL_Tree/AVL_Tree.cpp [hunk 14]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         cout<<ptr->data;
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after -> and <<

File: AVL_Tree/AVL_Tree.cpp [hunk 15]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         display(ptr->left, level + 1);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep a space before and after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 16]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
- * Inorder Traversal of AVL Tree
-
- */
-
 void avlTree::inorder(avl_node *tree)
-
 {
-
     if (tree == NULL)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the space after if

File: AVL_Tree/AVL_Tree.cpp [hunk 17]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         return;
-
     inorder (tree->left);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the space after the function name
keep a space before and after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 18]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
     cout<<tree->data<<"  ";
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep space before and after operators

File: AVL_Tree/AVL_Tree.cpp [hunk 19]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
     inorder (tree->right);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the extra space 
keep space after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 20]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
- * Preorder Traversal of AVL Tree
-
- */
-
 void avlTree::preorder(avl_node *tree)
-
 {
-
     if (tree == NULL)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the extra space after if

File: AVL_Tree/AVL_Tree.cpp [hunk 21]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         return;
-
     cout<<tree->data<<"  ";
-
     preorder (tree->left);
-
     preorder (tree->right);
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep space before and after -> and <<

File: AVL_Tree/AVL_Tree.cpp [hunk 22]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
- * Postorder Traversal of AVL Tree
-
- */
-
 void avlTree::postorder(avl_node *tree)
-
 {
-
     if (tree == NULL)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the extra space

File: AVL_Tree/AVL_Tree.cpp [hunk 23]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         return;
-
     postorder ( tree ->left );
-
     postorder ( tree ->right );
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the extra space 
keep space after ->

File: AVL_Tree/AVL_Tree.cpp [hunk 24]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
     cout<<tree->data<<"  ";
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: keep space after -> and <<

File: AVL_Tree/AVL_Tree.cpp [hunk 25]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
         display(ptr->right, level + 1);
-
         printf("\n");
-
         if (ptr == root)
-
         cout<<"Root -> ";
-
         for (i = 0; i < level && ptr != root; i++)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the space before parenthesis

File: AVL_Tree/AVL_Tree.cpp [hunk 26]
```diff
@@ -425,129 +372,54 @@ avl_node *avlTree::insert(avl_node *root, int value)
...
-
     if (ptr!=NULL)
-
     {
-
         display(ptr->right, level + 1);
-
         printf("\n");
-
         if (ptr == root)
```
Comment (asha15) [inline AVL_Tree/AVL_Tree.cpp]: remove the space after if

**Comment by jainaman224** [comment]
@ajaynair710 Please resolve conflicts.

**Comment by jainaman224** [comment]
@ajaynair710 Are you working on it?

**Comment by ajaynair710** [comment]
yes

**Comment by raghav-dalmia** [comment]
Do make all the changes. Also, squash your commits.

**Comment by somya-kapoor** [comment]
@ajaynair710 Are you working on it?

File: AVL_Tree/AVL_Tree.cpp (hunk 1 above)
Comment (somya-kapoor) [reply AVL_Tree/AVL_Tree.cpp:2]: remove this line

File: Boyer_Moore_Algorithm/Boyer_Moore.java [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+{
+    public void findPattern(String t, String p)
+    {
+        char[] text = t.toCharArray();
+        char[] pattern = p.toCharArray();
+        int pos = indexOf(text, pattern);
+        if (pos == -1)
+            System.out.println("\nNo Match\n");
+        else
+            System.out.println("Pattern found at position : "+ pos);
```
Comment (somya-kapoor) [inline Boyer_Moore_Algorithm/Boyer_Moore.java:16]: Add space before +

File: Boyer_Moore_Algorithm/Boyer_Moore.java [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+        if (pattern.length == 0) 
+            return 0;
+        int charTable[] = makeCharTable(pattern);
+        int offsetTable[] = makeOffsetTable(pattern);
+        for (int i = pattern.length - 1, j; i < text.length;) 
+        {
+            for (j = pattern.length - 1; pattern[j] == text[i]; --i, --j) 
+                     if (j == 0) 
+                    return i;
+
```
Comment (somya-kapoor) [inline Boyer_Moore_Algorithm/Boyer_Moore.java:31]: Remove extra lines

File: Boyer_Moore_Algorithm/Boyer_Moore.java [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+ 
+
+              // i += pattern.length - j; // For naive method
+              i += Math.max(offsetTable[pattern.length - 1 - j], charTable[text[i]]);
```
Comment (somya-kapoor) [inline Boyer_Moore_Algorithm/Boyer_Moore.java:35]: inconsistency in indentation

File: Boyer_Moore_Algorithm/Boyer_Moore.java [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        return -1;
```
Comment (somya-kapoor) [inline Boyer_Moore_Algorithm/Boyer_Moore.java:37]: inconsistency in indentation

File: Boyer_Moore_Algorithm/Boyer_Moore.java [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+      }
+
+      /** Makes the jump table based on the mismatched character information **/
+      private int[] makeCharTable(char[] pattern) 
+      {
+        final int ALPHABET_SIZE = 256;
```
Comment (somya-kapoor) [inline Boyer_Moore_Algorithm/Boyer_Moore.java:43]: inconsistency in indentation

File: Boyer_Moore_Algorithm/Boyer_Moore.java [hunk 32]
```diff
@@ -0,0 +1,100 @@
...
+               table[i] = pattern.length;
+        for (int i = 0; i < pattern.length - 1; ++i) 
+               table[pattern[i]] = pattern.length - 1 - i;
+        return table;
+      }
+
+      /** Makes the jump table based on the scan offset which mismatch occurs. **/
+      private static int[] makeOffsetTable(char[] pattern) 
+      {
+        int[] table = new int[pattern.length];
```
Comment (somya-kapoor) [inline Boyer_Moore_Algorithm/Boyer_Moore.java:55]: inconsistency in indentation

File: AVL_Tree/AVL_Tree.cpp [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        root = remove(x, root);
+    }
+
+    void display()
+    {
+        inorder(root);
+        cout << endl;
+    }
+};
```
Comment (somya-kapoor) [inline AVL_Tree/AVL_Tree.cpp:218]: Add sample

**Comment by somya-kapoor** [comment]
@ajaynair710 Are you working on it?

**Comment by somya-kapoor** [comment]
Issues #1666 and #1586 were already present.

**Pull request closed without merging**